File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/command/GetTripsByUidCommand.java
package com.worldventures.dreamtrips.modules.trips.command;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.R;
import com.worldventures.dreamtrips.api.trip.GetTripsDetailsHttpAction;
import com.worldventures.dreamtrips.api.trip.model.TripWithoutDetails;
import com.worldventures.dreamtrips.core.api.action.CommandWithError;
import com.worldventures.dreamtrips.core.janet.CommandActionBaseHelper;
import com.worldventures.dreamtrips.core.janet.JanetModule;
import com.worldventures.dreamtrips.core.janet.cache.CacheBundle;
import com.worldventures.dreamtrips.core.janet.cache.CacheBundleImpl;
import com.worldventures.dreamtrips.core.janet.cache.CacheOptions;
import com.worldventures.dreamtrips.core.janet.cache.CachedAction;
import com.worldventures.dreamtrips.core.janet.cache.ImmutableCacheOptions;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.modules.trips.model.TripModel;
import com.worldventures.dreamtrips.modules.trips.storage.TripsByUidsStorage;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import io.techery.janet.ActionHolder;
import io.techery.janet.Janet;
import io.techery.janet.command.annotations.CommandAction;
import io.techery.mappery.MapperyContext;

@CommandAction
public class GetTripsByUidCommand extends CommandWithError<List<TripModel>> implements InjectableAction,
      CachedAction<List<TripModel>> {

   @Inject @Named(JanetModule.JANET_API_LIB) Janet janet;
   @Inject MapperyContext mappery;

   private List<String> tripUids;

   private List<TripModel> cachedTrips;

   public GetTripsByUidCommand(List<String> tripUids) {
      this.tripUids = tripUids;
   }

   @Override
   protected void run(CommandCallback<List<TripModel>> callback) throws Throwable {
      if (hasValidCachedItems()) callback.onProgress(0);
      //
      janet.createPipe(GetTripsDetailsHttpAction.class)
            .createObservableResult(new GetTripsDetailsHttpAction(tripUids))
            .map(GetTripsDetailsHttpAction::response)
            .map(this::mapItems)
            .subscribe(CommandActionBaseHelper.ActionCommandSubscriber.wrap(callback));
   }

   public boolean hasValidCachedItems() {
      return cachedTrips != null && cachedTrips.size() > 0 && requestedSizeEqualsCached();
   }

   private boolean requestedSizeEqualsCached() {
      return cachedTrips.size() == tripUids.size();
   }

   private List<TripModel> mapItems(List<TripWithoutDetails> apiTrips) {
      return mappery.convert(apiTrips, TripModel.class);
   }

   public List<TripModel> getItems() {
      if (getResult() != null) return getResult();
      else return cachedTrips;
   }

   @Override
   public int getFallbackErrorMessage() {
      return R.string.string_failed_to_load_trips;
   }

   @Override
   public List<TripModel> getCacheData() {
      return new ArrayList<>(getResult());
   }

   @Override
   public void onRestore(ActionHolder holder, List<TripModel> cache) {
      cachedTrips = cache;
   }

   @Override
   public CacheOptions getCacheOptions() {
      CacheBundle cacheBundle = new CacheBundleImpl();
      cacheBundle.put(TripsByUidsStorage.UIDS, tripUids);
      return ImmutableCacheOptions.builder().params(cacheBundle).build();
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/service/command/GetTripsByUidCommand.kt
package com.worldventures.dreamtrips.modules.trips.service.command

import com.worldventures.core.janet.CommandWithError
import com.worldventures.dreamtrips.R
import com.worldventures.dreamtrips.api.trip.GetTripsDetailsHttpAction
import com.worldventures.dreamtrips.core.janet.CommandActionBaseHelper
import com.worldventures.dreamtrips.modules.trips.model.TripModel
import com.worldventures.dreamtrips.modules.trips.service.storage.TripsByUidsStorage
import com.worldventures.janet.cache.CacheBundleImpl
import com.worldventures.janet.cache.CacheOptions
import com.worldventures.janet.cache.CachedAction
import com.worldventures.janet.injection.InjectableAction
import io.techery.janet.ActionHolder
import io.techery.janet.Command
import io.techery.janet.Janet
import io.techery.janet.command.annotations.CommandAction
import io.techery.mappery.MapperyContext
import java.util.ArrayList
import javax.inject.Inject

@CommandAction
class GetTripsByUidCommand(private val tripUids: List<String>)
   : CommandWithError<List<TripModel>>(), InjectableAction, CachedAction<List<TripModel>> {

   @Inject internal lateinit var janet: Janet
   @Inject internal lateinit var mappery: MapperyContext

   private var cachedTrips: List<TripModel> = emptyList()

   val items: List<TripModel>
      get() = result ?: cachedTrips

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<List<TripModel>>) {
      if (hasValidCachedItems()) callback.onProgress(0)
      janet.createPipe(GetTripsDetailsHttpAction::class.java)
            .createObservableResult(GetTripsDetailsHttpAction(tripUids))
            .map { it.response() }
            .map { mappery.convert(it, TripModel::class.java) }
            .subscribe(CommandActionBaseHelper.ActionCommandSubscriber.wrap(callback))
   }

   fun hasValidCachedItems() = cachedTrips.isNotEmpty() && requestedSizeEqualsCached()

   private fun requestedSizeEqualsCached() = cachedTrips.size == tripUids.size

   override fun getFallbackErrorMessage() = R.string.string_failed_to_load_trips

   override fun getCacheData() = ArrayList(result)

   override fun onRestore(holder: ActionHolder<*>, cache: List<TripModel>) {
      cachedTrips = cache
   }

   override fun getCacheOptions() = CacheOptions(params = CacheBundleImpl().apply {
      put(TripsByUidsStorage.TRIP_UIDS, tripUids)
   })
}
